1037 intel compiler fixes for lfric examples

hiker opened the pull request:
I've fixed all LFRic compilation issues that I noticed with the Intel 20 compiler. I checked all examples for gocean, nemo, lfric. Running gocean did not work due to https://github.com/stfc/dl_esm_inf/issues/57 (when I fixed that bug as indicated in the ticket, the codes actuall ran).

hiker commented:
OK, I also fixed the compilation of the LFRic tutorials.

TeranIvy commented:
Happy to take it as we have an Intel compiler in hand 😄

hiker commented:
> Happy to take it as we have an Intel compiler in hand 😄

Perfect, since I had to modify your LFRic tutorials as well!

TeranIvy commented:
Hi @hiker! I have some issues with testing the full example with the Intel compiler due to recent parsing changes on master. Would you mind updating your branch? Also, could you please update the `README`s in the examples and the LFRic infrastructure library with the examples of how to set `F90` and `F90FLAGS` for the Intel compiler?

hiker commented:
> Hi @hiker! I have some issues with testing the full example with the Intel compiler due to recent parsing changes on master. Would you mind updating your branch? Also, could you please update the `README`s in the examples and the LFRic infrastructure library with the examples of how to set `F90` and `F90FLAGS` for the Intel compiler?

*grin* You are all working too hard - I updated it like 2 days ago, and it's already too old ;) OK, updated again. I've added Intel instructions to the infrastructure library, and added a new README to .../examples, which mentions it. I don't think we need to add this to all examples.

TeranIvy commented on lib/README.md:
```diff
@@ -22,6 +22,11 @@ rely on a file compiled in this directory.
+## `nan_test`
+
+Contains PSyData libraries for checking that input and output parameters of kernels
+are valid number (i.e. not NAN or infinite).
```
"are valid numbers". Also "NaN" and "infinity" (or just "Inf").

hiker replied:
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/README.md:
```diff
@@ -33,3 +33,11 @@ will add support for reading in meshes using NetCDF in a library
 have to link in NetCDF after ``liblfric_netcdf.a``.
+
+## Using the Intel Compiler
+In order to use the Intel compiler, use the F90 and F90FLAGS
```
Again, I'd suggest using code quotes for the compiler flags, ``F90`` and ``F90FLAGS``.

hiker replied:
Done.

TeranIvy replied:
It doesn't seem so :smiley:

TeranIvy replied:
Not quite :smiley:

hiker replied:
Same issue, push didn't come through.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/README.md:
````diff
@@ -33,3 +33,11 @@ will add support for reading in meshes using NetCDF in a library
+environmnent variables, e.g.:
+
+```shell
+F90=ifort F90FLAGS="-check bounds" make standalone
````
How come the ``-g`` option is not added here, whereas it is added in the ``examples`` README?

hiker replied:
I forgot 😸

TeranIvy commented on tutorial/practicals/LFRic/building_code/2_built_ins/builtins_driver.f90:
```diff
@@ -54,8 +54,8 @@ program builtins_driver
   ! Global and local mesh data types
-  type(global_mesh_type), target  :: global_mesh
-  type(global_mesh_type), pointer :: global_mesh_ptr => null()
+  type(global_mesh_base_type), target  :: global_mesh
```
Sorry to be nitpicky, but could you please align the `type` declarations by ``::` as they were before this change? :smiley:

hiker replied:
Done.

TeranIvy reviewed (changes requested):
The changes in general look good. Thank you for updating the branch and adding insructions for the Fortran compiler. 
I have some small style remarks (see comments inline) and a bigger problem, and that is runtime error I get in the LFRic infrastructure library with our Intel 17.0.1 compiler when I build the code with ``F90=ifort F90FLAGS="-g -check bounds" make`` as recommended:
```
forrtl: severe (408): fort: (3): Subscript #2 of the array CELL_NEXT_2D has value -1 which is less than the lower bound of 1

Image              PC                Routine            Line        Source             
example            0000000000608A06  Unknown               Unknown  Unknown
example            0000000000437DEE  Unknown               Unknown  Unknown
example            00000000004E38DB  Unknown               Unknown  Unknown
example            00000000004E1EE2  Unknown               Unknown  Unknown
example            00000000004D9359  Unknown               Unknown  Unknown
example            000000000040B70E  Unknown               Unknown  Unknown
example            0000000000402F5E  Unknown               Unknown  Unknown
libc-2.17.so       00007FA84CC9E555  __libc_start_main     Unknown  Unknown
example            0000000000402E69  Unknown               Unknown  Unknown
```
Grepping through the `src/psyclone/tests/test_files/dynamo0p3/infrastructure` code reveals that this array is allocated/deallocated/updated in the following source files:
- ``global_mesh_base_mod.F90``
- ``global_mesh_netcdf_mod.f90``
- ``mesh_mod.F90``
Could you please check what is happening there? At this point I'm not quite sure whether the bug is soemthing originally found in the LFRic code or something that was introduced by adapting the LFRic code to be used in PSyclone. If it is the former I'm happy to proceed without "check bounds" option for the Intel compiler for now and fix this in LFRic. If it is latter then it should be fixed in this PR.

hiker commented:
> I have some small style remarks (see comments inline) and a bigger problem, and that is runtime error I get in the LFRic infrastructure library with our Intel 17.0.1 compiler when I build the `examples/lfric/full_example` using `F90=ifort F90FLAGS="-g -check bounds" make` as recommended:
> 
> ```
> forrtl: severe (408): fort: (3): Subscript #2 of the array CELL_NEXT_2D has value -1 which is less than the lower bound of 1
> 
> Image              PC                Routine            Line        Source             
> example            0000000000608A06  Unknown               Unknown  Unknown
> example            0000000000437DEE  Unknown               Unknown  Unknown
> example            00000000004E38DB  Unknown               Unknown  Unknown
> example            00000000004E1EE2  Unknown               Unknown  Unknown
> example            00000000004D9359  Unknown               Unknown  Unknown
> example            000000000040B70E  Unknown               Unknown  Unknown
> example            0000000000402F5E  Unknown               Unknown  Unknown
> libc-2.17.so       00007FA84CC9E555  __libc_start_main     Unknown  Unknown
> example            0000000000402E69  Unknown               Unknown  Unknown
> ```

I believe this is a potential bug in LFRic: here the call sequence:
```
        call global_mesh%get_cell_next(cell_next(E), cell_next)
...
  subroutine get_cell_next (self, cell_gid, cell_next)
...
    cell_next(:) = self%cell_next_2d(:,cell_gid)
```
As you can cee, the integer variable ``cell_gid`` is ``cell_next(E)``. So when the assignment:
```
cell_next(:) = self%cell_next_2d(:,cell_grid)
```
is done, the value of ``cell_grid`` is actually changed. ``E`` is 3, and ``self%cell_next_2d`` is ``  2          -1          -1           6``. So the array assignment copies first 2, then -1. But the next -1  (third value) is stored in ``cell_next(E)``, which is ``cell_grid``. So when trying to copy the 4th value, ``cell_gid`` is now -1.
If you don't use debugging, this problem is avoided since it will create more efficient code. It can also be avoided by using e.g.:
```
    integer (i_def)                :: dummy(4)
    dummy = self%cell_next_2d(:,cell_gid)
    cell_next(:) = dummy
```
or in the calling subroutine use:
```
        dummy = cell_next(E)
        call global_mesh%get_cell_next(dummy, cell_next)
```
I still need to go through the rest of your review :)

hiker commented on examples/README.md:
```diff
@@ -0,0 +1,41 @@
+Various examples related to the NEMO API.
+
+## [lfric](./lfric)
+Various examples related to the LFric API.
```
Done.

TeranIvy replied:
It doesn't seem done :smiley:

TeranIvy replied:
Not quite :smiley:

hiker replied:
Sorry, the push apparently didn't get through :(

hiker commented on examples/README.md:
```diff
@@ -0,0 +1,41 @@
+included in PSyclone, which helps generating a Kernel's argument
+list and its datatypes from the Kernel metadata.
+
+## transformations
```
Yes, but:
1) transformation has no README.md that I could link to
2) I have no idea why this one example is here and not in the LFRIc dir ... though that could be because
3) it is  actually still dynamo0.1

In order to progress this PR, I have opened #1100 for this.

TeranIvy replied:
Thank you!

hiker commented on examples/README.md:
```diff
@@ -0,0 +1,41 @@
+# Compiler Settings
+By default, all examples are using ``gfortran`` as compiler.
+If you want to change the compiler, use the environment
+variables F90 and F90FLAGS, e.g.:
```
Done.

hiker commented:
Re the crash: I checked current LFRic, and most similar loops are correctly using a separate scalar value, only this one was not fixed. I've opened https://code.metoffice.gov.uk/trac/lfric/ticket/2512 to inform the LFRic team (including the simple fix). TBH, I can't remember enough about SVN to submit a PR - PLEEEEASE SWITCH TO GIT ... ah well, yes, I know the answer for that :cry: But at least I'll make them suffer 👿 

Updated to latest master, all issues addressed - ready for next review.

TeranIvy reviewed (changes requested):
Thanks for fixing the LFRic infrastructure bug in this PR and opening the LFRic ticket (which I am very happy to take on and provide the fix on the LFRic side).

I have checked the modified examples and tutorials with the Intel 17.0.1 compiler using the recommended flags and I am happy to report that things mostly work. There are a couple of exceptions (see points below) and forgotten changes from the first review (see inline comments).

- In the `full_example_netcdf` I successfully built the executable `example` but could not really run it as it was just hanging and not doing anything after running it with `./example`. This also happens with the Gnu compiler. Was I supposed to give the program an input parameter or is it something else? Could you please investigate?
- I think the drivers `simple_kernels_driver.f90` in `tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/` and `part2` also need to be updated.

hiker commented:
> Thanks for fixing the LFRic infrastructure bug in this PR and opening the LFRic ticket (which I am very happy to take on and provide the fix on the LFRic side).

It worked for me ... but that's not good enough. OK, I broke my bug fix in my final adjustment, please don't apply the code in the ticket yet. Oh, and I apparently forgot to push my final changes. Please stay tuned.

hiker commented:
> Thanks for fixing the LFRic infrastructure bug in this PR and opening the LFRic ticket (which I am very happy to take on and provide the fix on the LFRic side).

I've just updated it with the proper bug fix. I am not sure why this 'fix' originally worked for me, but I didn't notice that this particular loop already used the scalar variable ``cell`` (the others didn't) - so when I tried to bring it in sync with the other code, I broke it (even though I did compile and test ... no idea, maybe I forgot debug flag???)

> I have checked the modified examples and tutorials with the Intel 17.0.1 compiler using the recommended flags and I am happy to report that things mostly work. There are a couple of exceptions (see points below) and forgotten changes from the first review (see inline comments).
> 
> * In the `full_example_netcdf` I successfully built the executable `example` but could not really run it as it was just hanging and not doing anything after running it with `./example`. This also happens with the Gnu compiler. Was I supposed to give the program an input parameter or is it something else? Could you please investigate?

Good catch. Unfortunately I don't have a working netcdf with Intel on my machine, so I couldn't test it. I've just transferred this branch to a machine with intel (and netcdf etc), and it works for me now.

> * I think the drivers `simple_kernels_driver.f90` in `tutorial/practicals/LFRic/building_code/1_simple_kernels/part1/` and `part2` also need to be updated.

Good catch. I also re-checked the other LFRic tutorials, they seem fine now.

TeranIvy reviewed (approved):
All well now, thanks. Proceeding to merge.
